crates/ruma-macros/src/events/event_type.rs

```diff
@@ -86,21 +85,70 @@ fn generate_enum(
     }
 
     let event_types = deduped.iter().map(|e| &e.ev_type);
-    let variants =
-        deduped.iter().map(|e| Ok(e.to_variant()?.decl())).collect::<syn::Result<Vec<_>>>()?;
+
+    let variants: Vec<_> = deduped
+        .iter()
+        .map(|e| {
+            let start = e.to_variant()?.decl();
+            let data = e.has_type_fragment().then(|| quote! { (::std::string::String) });
+
+            Ok(quote! {
+                #start #data
+            })
+        })
+        .collect::<syn::Result<_>>()?;
+
+    let to_cow_str_match_arms: Vec<_> = deduped
+        .iter()
+        .map(|e| {
+            let v = e.to_variant()?;
+            let start = v.match_arm(quote! { Self });
+            let ev_type = &e.ev_type;
+
+            Ok(if let Some(prefix) = ev_type.value().strip_suffix(".*") {
+                let fstr = prefix.to_owned() + "{}";
+                quote! { #start(_s) => ::std::borrow::Cow::Owned(::std::format!(#fstr, _s)) }
+            } else {
+                quote! { #start => ::std::borrow::Cow::Borrowed(#ev_type) }
+            })
+        })
+        .collect::<syn::Result<_>>()?;
+
+    let from_str_match_arms: Vec<_> = deduped
+        .iter()
+        .map(|e| {
+            let v = e.to_variant()?;
+            let ctor = v.ctor(quote! { Self });
+
+            let match_arm = if let Some(prefix) = e.ev_type.value().strip_suffix(".*") {
+                let match_prefix = format!("{}.", prefix);
```

Comment:
```suggestion
            let match_arm = if let Some(match_prefix) = e.ev_type.value().strip_suffix("*") {
```
This would work to right?

Reply:
I guess it would because we check elsewhere that `*` only exists as part of a `.*` suffix.